README.md
<!-- An easy-to-use and simple Instagram package allows you to fetch media content without API and access token -->

<p align="center">
  <img width="50%" src="https://raw.githubusercontent.com/webistomin/nanogram.js/260d58af351d84606a75b32160f349c9a444691d/assets/logo.svg" alt="Nanogram">
</p>

---

<h4 align="center">An easy-to-use and simple Instagram package allows you to fetch media content <strong>without API</strong> and <strong>access token</strong>.</h4>

<h5 align="center">⭐️ Star me on GitHub — it helps!</h5>

<p align="center">
  <a href="https://travis-ci.org/webistomin/nanogram.js">
    <img src="https://travis-ci.org/webistomin/nanogram.js.svg?branch=master"
         alt="Travis CI">
  </a>
  <a href="https://codecov.io/gh/webistomin/nanogram.js">
    <img src="https://codecov.io/gh/webistomin/nanogram.js/branch/master/graph/badge.svg" alt="codecoverage" />
  </a>
  <a href="https://www.codefactor.io/repository/github/webistomin/nanogram.js"><img src="https://www.codefactor.io/repository/github/webistomin/nanogram.js/badge" alt="CodeFactor" /></a>
  <img alt="GitHub issues" src="https://img.shields.io/github/issues/webistomin/nanogram.js">
  <img alt="GitHub last commit" src="https://img.shields.io/github/last-commit/webistomin/nanogram.js">
  <img alt="npm type definitions" src="https://img.shields.io/npm/types/nanogram.js">
  <a href="https://bundlephobia.com/result?p=nanogram.js@3.0.2">
  <img alt="npm bundle size" src="https://img.shields.io/bundlephobia/minzip/nanogram.js/3.0.2">
  </a>
</p>

<p align="center">
<a href="https://www.npmjs.com/package/nanogram.js"><img alt="npm" src="https://img.shields.io/npm/v/nanogram.js" /></a>
<a href="https://www.npmjs.com/package/nanogram.js"><img alt="downloads" src="https://img.shields.io/npm/dm/nanogram.js" /></a>
  <a href="https://www.jsdelivr.com/package/npm/nanogram.js"><img alt="jsdeliver" src="https://data.jsdelivr.com/v1/package/npm/nanogram.js/badge?style=rounded" /></a>
<a href="#contributors-"><img alt="contributors" src="https://img.shields.io/badge/all_contributors-2-orange.svg" /></a>
</p>

<p align="center">
  <a href="#key-features-">Key Features</a> •
  <a href="#demo-">Demo</a> •
  <a href="#documentation-">Documentation</a> •
  <a href="#installation-">Installation</a> •
  <a href="#how-to-use-">How To Use</a> •
  <a href="#caveats-">Caveats</a> •
  <a href="#browsers-support-">Browsers support</a> •
  <a href="#license-">License</a>  •
  <a href="#contributing-">Contributing</a>
</p>

## Key Features ✨

* **Small.** ~ 2KB (minified and gzipped). <a href="https://github.com/ai/size-limit">Size Limit</a> controls the size
* **No dependencies**
* No need for the **access token** secret
* **Easy to use**
* **Typescript** support
* **Tree shakeable**

## Demo 👀

### [DEMO](https://codepen.io/webistomin/pen/vYNrgYL)

## Documentation 🔨

You can read the full documentation [here](https://nanogram-js.now.sh/)

## Installation 🚀

### Using package managers

#### npm
```shell script
$ npm install nanogram.js --save
```

#### yarn
```shell script
$ yarn add nanogram.js
```

### via CDN

Add script right before closing `</body>` tag

```html
<script src="https://unpkg.com/nanogram.js@3.0.2/dist/nanogram.iife.js"></script>
or
<script src="https://cdn.jsdelivr.net/npm/nanogram.js@3.0.2/dist/nanogram.iife.js"></script>
```

_Hint:_ for a better performance add [preconnect](https://css-tricks.com/using-relpreconnect-to-establish-network-connections-early-and-increase-performance/) link in the head of your document. 

```html 
<head>
  <!-- for unkpg cdn --> 
  <link rel="preconnect" href="https://unpkg.com">

  <!-- for jsdelivr cdn -->
  <link rel="preconnect" href="https://cdn.jsdelivr.net">  


  <!-- dns-prefetch only for IE11 --> 
  <!--	<link rel="dns-prefetch" href="https://unpkg.com"> -->
  <!--	<link rel="dns-prefetch" href="https://cdn.jsdelivr.net"> -->
</head>
```

## How to use 🤔

### **Get media content by providing instagram username**

#### ES2015

```js
import { getMediaByUsername } from 'nanogram.js'

getMediaByUsername('instagram').then((media) => {
  console.log(media);
});
```

#### CommonJS

```js
const getMediaByUsername = require('nanogram.js').getMediaByUsername;

getMediaByUsername('instagram').then((media) => {
  console.log(media);
});
```

#### IIFE

```js
const getMediaByUsername = window.Nanogram.getMediaByUsername

getMediaByUsername('instagram').then((media) => {
  console.log(media);
});
```

_Note:_ get content from [user page](https://www.instagram.com/instagram/). 12 photos is the maximum for this method.

---

### **Get media content by providing instagram post id**

#### ES2015

```js
import { getMediaByPostId } from 'nanogram.js'

getMediaByPostId('CIrIDMtDwn4').then((media) => {
  console.log(media);
});
```

#### CommonJS

```js
const getMediaByPostId = require('nanogram.js').getMediaByPostId;

getMediaByPostId('CIrIDMtDwn4').then((media) => {
  console.log(media);
});
```

#### IIFE

```js
const getMediaByPostId = window.Nanogram.getMediaByPostId

getMediaByPostId('CIrIDMtDwn4').then((media) => {
  console.log(media);
});
```

_Note_: get content from [post page](https://www.instagram.com/p/CIrIDMtDwn4/)

---

### **Get media content by providing instagram tag**

#### ES2015

```js
import { getMediaByTag } from 'nanogram.js'

getMediaByUsername('sunset').then((media) => {
  console.log(media);
});
```

#### CommonJS

```js
const getMediaByTag = require('nanogram.js').getMediaByTag;

getMediaByUsername('sunset').then((media) => {
  console.log(media);
});
```

#### IIFE

```js
const getMediaByTag = window.Nanogram.getMediaByTag

getMediaByTag('sunset').then((media) => {
  console.log(media);
});
```

_Note:_ get content from [tag page](https://www.instagram.com/explore/tags/sunset/)

---

### **Get media content by providing location id and place name**

#### ES2015

```js
import { getMediaByLocation } from 'nanogram.js'

getMediaByLocation(6264386, 'highbridge-park').then((media) => {
  console.log(media);
});
```

#### CommonJS

```js
const getMediaByLocation = require('nanogram.js').getMediaByLocation;

getMediaByLocation(6264386, 'highbridge-park').then((media) => {
  console.log(media);
});
```

#### IIFE

```js
const getMediaByLocation = window.Nanogram.getMediaByLocation

getMediaByLocation(6264386, 'highbridge-park').then((media) => {
  console.log(media);
});
```

_Note:_ get content from [location page](https://www.instagram.com/explore/locations/6264386/highbridge-park)

---

### **Get all available countries**

#### ES2015

```js
import { getCountries } from 'nanogram.js'

getCountries().then((media) => {
  console.log(media);
});
```

#### CommonJS

```js
const getCountries = require('nanogram.js').getCountries;

getCountries().then((media) => {
  console.log(media);
});
```

#### IIFE

```js
const getCountries = window.Nanogram.getCountries

getCountries().then((media) => {
  console.log(media);
});
```

_Note:_ get countries from [location page](https://www.instagram.com/explore/locations/)

---

### **Get all cities by providing country id**

#### ES2015

```js
import { getCitiesByCountryId } from 'nanogram.js'

getCitiesByCountryId('US').then((media) => {
  console.log(media);
});
```

#### CommonJS

```js
const getCitiesByCountryId = require('nanogram.js').getCitiesByCountryId;

getCitiesByCountryId('US').then((media) => {
  console.log(media);
});
```

#### IIFE

```js
const getCitiesByCountryId = window.Nanogram.getCitiesByCountryId

getCitiesByCountryId('US').then((media) => {
  console.log(media);
});
```

_Note:_ get cities from [country page](https://www.instagram.com/explore/locations/US/)

---

### **Get all places by providing city id**

#### ES2015

```js
import { getPlacesByCityId } from 'nanogram.js'

getPlacesByCityId('c2728325').then((media) => {
  console.log(media);
});
```

#### CommonJS

```js
const getPlacesByCityId = require('nanogram.js').getPlacesByCityId;

getPlacesByCityId('c2728325').then((media) => {
  console.log(media);
});
```

#### IIFE

```js
const getPlacesByCityId = window.Nanogram.getPlacesByCityId

getPlacesByCityId('c2728325').then((media) => {
  console.log(media);
});
```

_Note:_ get places from [city page](https://www.instagram.com/explore/locations/c2728325/)

---

### **Get media content by providing place id**

#### ES2015

```js
import { getMediaByPlaceId } from 'nanogram.js'

getMediaByPlaceId(2999512).then((media) => {
  console.log(media);
});
```

#### CommonJS

```js
const getMediaByPlaceId = require('nanogram.js').getMediaByPlaceId;

getMediaByPlaceId(2999512).then((media) => {
  console.log(media);
});
```

#### IIFE

```js
const getMediaByPlaceId = window.Nanogram.getMediaByPlaceId

getMediaByPlaceId(2999512).then((media) => {
  console.log(media);
});
```

_Note:_ get content from [place page](https://www.instagram.com/explore/locations/2999512/)

---

### **Get media content by providing search query**

#### ES2015

```js
import { getMediaBySearchQuery } from 'nanogram.js'

getMediaBySearchQuery('sunset').then((media) => {
  console.log(media);
});
```

#### CommonJS

```js
const getMediaBySearchQuery = require('nanogram.js').getMediaBySearchQuery;

getMediaBySearchQuery('sunset').then((media) => {
  console.log(media);
});
```

#### IIFE

```js
const getMediaBySearchQuery = window.Nanogram.getMediaBySearchQuery

getMediaBySearchQuery('sunset').then((media) => {
  console.log(media);
});
```

_Note:_ get content from searchbar on the top of the page

---

## Caveats 💣

<details>
    <summary>How to get next page? There are only 12 photos</summary>
    <br/>
    12 photos is the maximum. I think there is no way to load more content.
Use instafeed.js with access_token to load more items.
</details>

<details>
    <summary>Nanogram stopped working (Cannot read property '0' of undefined)</summary>
    <br/>
     Perhaps you made a large number of requests in a short period of time. There is a limit.
      Usually, the duration of a temporary Instagram ban ranges from few hours to 24-48 hours. The duration of ban also depends on your follow up actions. If you would continue doing the wrong actions, the ban may prolong.
</details>

## Browsers support 🌎

| [<img src="https://raw.githubusercontent.com/alrra/browser-logos/master/src/edge/edge_48x48.png" alt="IE / Edge" width="24px" height="24px" />](http://godban.github.io/browsers-support-badges/)<br/>IE / Edge | [<img src="https://raw.githubusercontent.com/alrra/browser-logos/master/src/firefox/firefox_48x48.png" alt="Firefox" width="24px" height="24px" />](http://godban.github.io/browsers-support-badges/)<br/>Firefox | [<img src="https://raw.githubusercontent.com/alrra/browser-logos/master/src/chrome/chrome_48x48.png" alt="Chrome" width="24px" height="24px" />](http://godban.github.io/browsers-support-badges/)<br/>Chrome | [<img src="https://raw.githubusercontent.com/alrra/browser-logos/master/src/safari/safari_48x48.png" alt="Safari" width="24px" height="24px" />](http://godban.github.io/browsers-support-badges/)<br/>Safari | [<img src="https://raw.githubusercontent.com/alrra/browser-logos/master/src/safari-ios/safari-ios_48x48.png" alt="iOS Safari" width="24px" height="24px" />](http://godban.github.io/browsers-support-badges/)<br/>iOS Safari | [<img src="https://raw.githubusercontent.com/alrra/browser-logos/master/src/opera/opera_48x48.png" alt="Opera" width="24px" height="24px" />](http://godban.github.io/browsers-support-badges/)<br/>Opera | [<img src="https://raw.githubusercontent.com/alrra/browser-logos/master/src/yandex/yandex_48x48.png" alt="Yandex" width="24px" height="24px" />](http://godban.github.io/browsers-support-badges/)<br/>Yandex |
| --------- | --------- | --------- | --------- | --------- | --------- | --------- |
| IE11*, Edge 12+| 27+ | 32+ | 7.1+| 8+ | 19+| 14.2+

`*` – For IE11 you need to install a **Promise** polyfill.

If you are using version **less than 2.0.0** you will also need polyfill for **Fetch API**.

If you want maximum browser compatibility, please use polyfills.

See caniuse for [promise](https://caniuse.com/#search=promise).

### Browsers polyfills

If you want to send a polyfill only to browsers that need it, there's a handy service called [Polyfill.io](https://polyfill.io/v3/url-builder/) which does just that, it offers a wide array of polyfills.

Here's an example of using [polyfill.io](https://polyfill.io/v3/url-builder/) to polyfill only the `Promise` feature, so if we put this right before closing `</body>` tag of `index.html` and `Nanogram.js` script, Polyfill.io will read the user agent and use that information to determine if the browser requires a polyfill for the feature or features listed. Since I'm using Chrome it will send back an empty response since my browser doesn't need it, pretty slick.

```html
<script src="https://polyfill.io/v3/polyfill.min.js?features=Promise"></script>
```

Alternatively, you can install packages like [es6-promise](https://github.com/stefanpenner/es6-promise).

### Node.js

Node.js doesn't have a built-in implementation of the `XHR API`, but you can use any library with a compatible interface, such [xmlhttprequest](https://www.npmjs.com/package/xmlhttprequest).

If you are using version **less than 2.0.0** you need `Fetch API` instead of `XHR API`. Such [node-fetch](https://github.com/node-fetch/node-fetch).

## License 📄

### [MIT](https://github.com/webistomin/nanogram.js/blob/master/LICENSE)

## Contributing 🎉

Found a bug? Missing a specific feature?
Your contributions are always welcome! Please have a look at the [contribution guidelines](https://github.com/webistomin/nanogram.js/blob/master/CONTRIBUTING.md) first.

## Contributors ✨

Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/docs/en/emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<table>
  <tr>
    <td align="center"><a href="https://webistom.in/en"><img src="https://avatars0.githubusercontent.com/u/30475699?v=4" width="100px;" alt=""/><br /><sub><b>Alexey Istomin</b></sub></a><br /><a href="#infra-webistomin" title="Infrastructure (Hosting, Build-Tools, etc)">🚇</a> <a href="https://github.com/webistomin/nanogram.js/commits?author=webistomin" title="Code">💻</a> <a href="#ideas-webistomin" title="Ideas, Planning, & Feedback">🤔</a></td>
    <td align="center"><a href="https://github.com/yusufkhan07"><img src="https://avatars1.githubusercontent.com/u/24753161?v=4" width="100px;" alt=""/><br /><sub><b>Yousuf Khan</b></sub></a><br /><a href="https://github.com/webistomin/nanogram.js/commits?author=yusufkhan07" title="Code">💻</a> <a href="https://github.com/webistomin/nanogram.js/commits?author=yusufkhan07" title="Tests">⚠️</a> <a href="https://github.com/webistomin/nanogram.js/issues?q=author%3Ayusufkhan07" title="Bug reports">🐛</a></td>
  </tr>
</table>

<!-- markdownlint-enable -->
<!-- prettier-ignore-end -->
<!-- ALL-CONTRIBUTORS-LIST:END -->

This project follows the [all-contributors](https://github.com/all-contributors/all-contributors) specification. Contributions of any kind welcome!


CHANGELOG.md
## 1.0.0 (20.04.2020)

- Added: Initial version

## 1.0.1 (10.05.2020)
- Reduce bundle size (-10%) so now it's 1.99 KB
- Remove test folder from npm package
- Add node.js example
- Update README.md

## 1.0.2 (25.05.2020)
- Fix [issue#14](https://github.com/webistomin/nanogram.js/issues/14)
- Remove some devDependencies
- Update devDependencies

## 2.0.0 (20.07.2020)
- Remove some devDependencies
- Update devDependencies
- Fix [issue#58](https://github.com/webistomin/nanogram.js/issues/58)
- Fix [issue#91](https://github.com/webistomin/nanogram.js/issues/91)
- Fetch replaced by XHR. Now you don't need polyfill for Fetch API. If you have any errors – open an issue.

## 2.0.1 (27.10.2020)
- Add named export for cjs bundle
- Update devDependencies

## 2.0.2 (27.10.2020)
- Add sideEffect: false to package.json

## 3.0.0 (26.12.2020)

- Improved tree-shaking
- Added new method `getMediaByPostId`
- `ok` field removed from response
- Reduced library size
- Fix [issue#198](https://github.com/webistomin/nanogram.js/issues/198)
- Fix [issue#228](https://github.com/webistomin/nanogram.js/issues/228)

## 3.0.1 (26.12.2020)

- Fixed dist folder in npm package



src/nanogram.ts
import { IUserProfileResponse, IUserProfileResult } from './types/user-profile-page';
import { ITagsResponse, ITagsResult } from './types/tags-page';
import { ILocationResponse, ILocationResult } from './types/location-page';
import { ILocationDirectoryResponse, ILocationDirectoryResult } from './types/location-directory-page';
import { ICitiesResponse, ICitiesResult } from './types/cities-page';
import { IPlacesResponse, IPlacesResult } from './types/places-page';
import { IPlaceResponse, IPlaceResult } from './types/place-page';
import { ISearchResponse, ISearchResult } from './types/search-page';

export default class Nanogram {
  private readonly INSTAGRAM_HOSTNAME: string;
  private readonly SHARED_DATA_TAG_EXP: RegExp;

  constructor() {
    /*
     * Instagram base url
     */
    this.INSTAGRAM_HOSTNAME = 'https://www.instagram.com/';
    /*
     * Instagram RegExp to parse <script type="text/javascript">window._sharedData = {...}</script>
     */
    this.SHARED_DATA_TAG_EXP = /^[\w\W]*<script type="text\/javascript">window._sharedData = ({[\w\W]*});<\/script>[\w\W]*$/g;
  }

  /*
   * Add query to instagram base url
   */
  private buildUrl(query: string): string {
    return `${this.INSTAGRAM_HOSTNAME}${query}`;
  }

  /*
   * Get javascript object from response
   */
  private parseJSON<T>(content: string, useRegExp: boolean): T {
    try {
      let data = content;

      if (useRegExp) {
        data = content.replace(this.SHARED_DATA_TAG_EXP, '$1');
      }

      return JSON.parse(data) as T;
    } catch (error) {
      console.error(`Nanogram: failure during parsing JSON.\nError message: ${(<Error>error).message}`);
    }
  }

  /*
   * Make fetch call to instagram
   */
  private async HTTP<T>(request: string, useRegExp = true): Promise<T | undefined> {
    const xhrrequest = (method: string, url: string): Promise<XMLHttpRequest> => {
      return new Promise((resolve, reject) => {
        const xhr = new XMLHttpRequest();
        xhr.open(method, url);
        xhr.onload = (): void => resolve(xhr);
        xhr.onerror = reject;
        xhr.send();
      });
    };

    const response = await xhrrequest('GET', request).then((response: XMLHttpRequest) => {
      if (response.status >= 200 && response.status < 400) {
        return response.responseText;
      } else {
        console.error(
          [
            'Nanogram: error during request',
            'Probably making too many requests to the Instagram application.',
            'Also check method parameters',
          ].join('\n')
        );
      }
    });

    if (response) {
      return this.parseJSON(response, useRegExp);
    }

    return;
  }

  /*
   * Build error string from arguments
   */
  private static logError(params: string[]): void {
    const message = `Nanogram: please provide a valid ${params.join(' and ')}`;
    console.error(message);
  }

  /*
   * Get content from
   * https://www.instagram.com/{username}/
   *
   * Example:
   * https://www.instagram.com/instagram/
   */
  public async getMediaByUsername(username: string): Promise<IUserProfileResult> {
    const result: IUserProfileResult = {
      profile: null,
      ok: false,
    };

    if (!username) {
      Nanogram.logError(['username']);
      return result;
    }

    const url = this.buildUrl(username);
    const response = await this.HTTP<IUserProfileResponse>(url);

    const profile = response?.entry_data?.ProfilePage[0]?.graphql?.user || null;

    return { ...result, ...{ profile, ok: Boolean(profile) } };
  }

  /*
   * Get content from
   * https://www.instagram.com/explore/tags/{tag}/
   *
   * Example:
   * https://www.instagram.com/explore/tags/sunset/
   */
  public async getMediaByTag(tag: string): Promise<ITagsResult> {
    const result: ITagsResult = {
      tag: null,
      ok: false,
    };

    if (!tag) {
      Nanogram.logError(['tag']);
      return result;
    }

    const url = this.buildUrl(`explore/tags/${tag}`);
    const response = await this.HTTP<ITagsResponse>(url);
    const hashtag = response?.entry_data?.TagPage[0]?.graphql.hashtag || null;

    return { ...result, ...{ tag: hashtag, ok: Boolean(hashtag) } };
  }

  /*
   * Get content from
   * https://www.instagram.com/explore/locations/{locationId}/{placeName}
   *
   * Example:
   * https://www.instagram.com/explore/locations/6264386/highbridge-park
   */
  public async getMediaByLocation(locationId: number, placeName: string): Promise<ILocationResult> {
    const result: ILocationResult = {
      location: null,
      ok: false,
    };

    if (locationId === undefined || !placeName) {
      Nanogram.logError(['location id', 'place name']);
      return result;
    }

    const url = this.buildUrl(`explore/locations/${locationId}/${placeName}`);
    const response = await this.HTTP<ILocationResponse>(url);
    const location = response?.entry_data?.LocationsPage[0]?.graphql?.location || null;

    return { ...result, ...{ location, ok: Boolean(location) } };
  }

  /*
   * Get content from
   * https://www.instagram.com/explore/locations/
   *
   * Example:
   * https://www.instagram.com/explore/locations/
   */
  public async getCountries(): Promise<ILocationDirectoryResult> {
    const result: ILocationDirectoryResult = {
      country_list: null,
      ok: false,
    };

    const url = this.buildUrl(`explore/locations/`);
    const response = await this.HTTP<ILocationDirectoryResponse>(url);
    const countryList = response?.entry_data?.LocationsDirectoryPage[0]?.country_list || null;

    return { ...result, ...{ country_list: countryList, ok: Boolean(countryList) } };
  }

  /*
   * Get content from
   * https://www.instagram.com/explore/locations/{countryId}/
   *
   * Example:
   * https://www.instagram.com/explore/locations/US/
   */
  public async getCitiesByCountryId(countryId: string): Promise<ICitiesResult> {
    const result: ICitiesResult = {
      city_list: null,
      country_info: null,
      ok: false,
    };

    if (!countryId) {
      Nanogram.logError(['country id']);
      return result;
    }

    const url = this.buildUrl(`explore/locations/${countryId}`);
    const response = await this.HTTP<ICitiesResponse>(url);
    const { city_list = null, country_info = null } = { ...response?.entry_data?.LocationsDirectoryPage[0] };

    return { ...result, ...{ country_info, city_list, ok: Boolean(country_info || city_list) } };
  }

  /*
   * Get content from
   * https://www.instagram.com/explore/locations/{cityId}/
   *
   * Example:
   * https://www.instagram.com/explore/locations/c2728325/
   */
  public async getPlacesByCityId(cityId: string): Promise<IPlacesResult> {
    const result: IPlacesResult = {
      place: {
        city_info: null,
        country_info: null,
        location_list: null,
      },
      ok: false,
    };

    if (!cityId) {
      Nanogram.logError(['city id']);
      return result;
    }

    const url = this.buildUrl(`explore/locations/${cityId}`);
    const response = await this.HTTP<IPlacesResponse>(url);
    const { city_info = null, country_info = null, location_list = null } = {
      ...response?.entry_data?.LocationsDirectoryPage[0],
    };

    return {
      ...result,
      ...{ place: { city_info, country_info, location_list }, ok: Boolean(city_info || country_info || location_list) },
    };
  }

  /*
   * Get content from
   * https://www.instagram.com/explore/locations/{placeId}/
   *
   * Example:
   * https://www.instagram.com/explore/locations/2999512/
   */
  public async getMediaByPlaceId(placeId: number): Promise<IPlaceResult> {
    const result: IPlaceResult = {
      location: null,
      ok: false,
    };

    if (!placeId) {
      Nanogram.logError(['place id']);
      return result;
    }

    const url = this.buildUrl(`explore/locations/${placeId}`);
    const response = await this.HTTP<IPlaceResponse>(url);
    const location = response?.entry_data?.LocationsPage[0]?.graphql.location || null;

    return { ...result, ...{ location, ok: Boolean(location) } };
  }

  /*
   * Get content from
   * https://www.instagram.com/web/search/topsearch/?context=blended&query={query}&include_reel=true
   *
   * Example:
   * https://www.instagram.com/web/search/topsearch/?context=blended&query=sunset&include_reel=true
   */
  public async getMediaBySearchQuery(query: string): Promise<ISearchResult> {
    const result: ISearchResult = {
      media: {
        users: null,
        hashtags: null,
        places: null,
      },
      ok: false,
    };

    if (!query) {
      Nanogram.logError(['search query']);
      return result;
    }

    const url = this.buildUrl(`web/search/topsearch/?context=blended&query=${query}&include_reel=true`);
    const { users = null, hashtags = null, places = null } = await this.HTTP<ISearchResponse>(url, false);

    return { ...result, ...{ media: { users, hashtags, places }, ok: Boolean(users || hashtags || places) } };
  }
}
